HSC mass to energy conversion

RuaridhMacd opened the pull request:
# Description

This PR updates and rebases Mary's previous work on converting the hydrogen supply chain (HSC) from mass to energy units, i.e. tonnes H2 to MWh of H2. The energy values can be lower heating value of higher heating value as long as everything is consistent across sectors.

This is a breaking change as the names of some inputs has been changed.

I have created a script to convert mass-based data files to energy-based ones. [It can be found here.](https://github.com/RuaridhMacd/Dolphyn-HSC-Mass-2-Energy)

I have run tests on all the systems using the default TDR settings and Gurobi.

The original objective values are:
<img width="515" alt="image" src="https://github.com/user-attachments/assets/c792688c-4673-4f65-93e6-ab76faa006d2">

The current objective values (using the same TDR mapping) are:
<img width="515" alt="image" src="https://github.com/user-attachments/assets/28f13b56-c1f6-48b5-95d9-327bb3131735">

As you can see, all of the results are the same except for the Eastern_USA example which has a $3 difference (~1e-12 error). This example uses the LFSC model and has some numerical stability issues, so I don't think there is an underlying issue with the model. However, I welcome others checking.

I have tried my best to update all documentation but would appreciate people checking more missed values.

Finally, I have not made this work with the biofuels sector. That will be done in future work.

## Fixes issue \#

Fixes #202 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

- [x] Example_Systems/SmallNewEngland/OneZone 
- [x] Example_Systems/SmallNewEngland/ThreeZones

**Test Configuration**:
* OS: MacOS
* Solver: Gurobi
* Julia version: 1.10.4
* Solver version: 10

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run tests with my code to avoid compatibility issues
- [x] Any dependent changes have been merged and published in downstream modules

bennettm committed:
Convert h2_gen from mass to energy unit

bennettm committed:
Replaced tonne_per_hour with MWh

bennettm committed:
Convert mass to energy based on HSC_generation.csv column names

bennettm committed:
Column names from tonne to MW

RuaridhMacd committed:
Update example system data

Co-Authored-By: bennettm <2174909+bennettm@users.noreply.github.com>

RuaridhMacd committed:
Update HSC truck and liquid H2 names and docstrings

Co-Authored-By: bennettm <2174909+bennettm@users.noreply.github.com>

RuaridhMacd committed:
Update hsc_mass_2_energy.jl

Co-Authored-By: bennettm <2174909+bennettm@users.noreply.github.com>

RuaridhMacd committed:
Update remaining HSC and LFSC files

Co-Authored-By: bennettm <2174909+bennettm@users.noreply.github.com>

RuaridhMacd committed:
Update examples

Co-Authored-By: bennettm <2174909+bennettm@users.noreply.github.com>

RuaridhMacd committed:
Update comments and doc strings

Co-Authored-By: bennettm <2174909+bennettm@users.noreply.github.com>

RuaridhMacd committed:
Update HSC documentation

Co-Authored-By: bennettm <2174909+bennettm@users.noreply.github.com>

RuaridhMacd commented:
Updates needed:

- [ ] Swap rate-based emissions cap to use HHV, not LHV
- [ ] Move conversation script into Dolphyn, but exclude from Dolphyn.jl
- [ ] Update TruckCap_MW_per_Unit to TruckCap_MWh_per_Unit in code
- [ ] Update TruckCap_MW_per_Unit to TruckCap_MWh_per_Unit in documentation

RuaridhMacd committed:
Change rate-based emissions constraint to HHV, from LHV

RuaridhMacd committed:
Update TruckCap_MW_per_unit to TruckCap_MWh_per_unit

RuaridhMacd committed:
Update hsc_data_documentation.md

RuaridhMacd commented:
@dharik13 I've made the updates we've talked about in the Dolphyn meeting.

I've moved the hsc_mass_2_energy.jl conversion script to the Example_Systems folder as it'll be easy to find and won't be included in the compiled code. I'm happy for you to re-organize or move it wherever you think is best.

In my testing, the Obj values for the example systems still match those using the develop branch. When testing it, I recommend:

- Run run_examples.jl on develop with `force_TDR_recluster = true`. This will give you baseline results
- Switch to the HSC-mass-to-energy branch
- Run hsc_mass_2_energy.jl to change the units of all the examples, including the TDR files
- Set force_TDR_recluster = false in run_examples.jl, to avoid the TDR periods changing
- Run run_examples.jl to get the updated results

dharik13 commented on Example_Systems/ERCOT_1stg_hourly_5GW_base_tmr/HSC_load_data.csv:
Suggest reducing Voll to be a smaller number with fewer significant digits (e.g. $50k/MWh)

RuaridhMacd replied:
I've updated it to be 6 significant figures, so we stay relatively close to the original data.
It may be worthwhile to refresh all the data in the examples later

dharik13 commented on Example_Systems/Eastern_US_CSC/ThreeZones/HSC_G2P.csv:
Would suggest reducing number of significant digits for etaG2P parameter

dharik13 committed:
Renamed TruckEnergy to TruckChargePower

Renamed variables, expressions and parameters related to phrase "TruckEnergy" with "TruckChargePower" that better reflects the physical meaning of these terms.

TruckChargePower (previously labeled as TruckEnergy) refers to the power requirement associated with filling up a truck and implicitly limits the rate of filling up of the truck.
If the truck is a gas --> TruckChargePower refers to compressor charging requirement
If the truck is a liquid H2 --> Truck ChargePower could refer to pumping requirement or even liquefaction capacity

dharik13 committed:
Updated examples and hsc_mass_2_energy to include ChargePower

Updated remaining files to update parameter names to using "Energy" to use "ChargePower" which is more intuitive representation of the parameter

dharik13 committed:
Update write_h2_cost

Changed name of eTotalCFixH2TruckNumberPower to eTotalCFixH2TruckChargePower . This expression corresponds to fixed cost of resource capacity that limits rate of charging of the truck.

dharik13 commented on src/multisector/genx_hsc/co2_cap_power_hsc.jl:
There is an issue with this file. We dont define eEmissionsConstraintRHS for the power sector for different cases of setup[CO2Cap] - see main version where it is done correctly. Something got messed up here.

RuaridhMacd replied:
- [x] To-do @RuaridhMacd

dharik13 committed:
Further edits on name change

Updated documentation to change the parameter names rleated to trucks and also updated the example files.

dharik13 reviewed (changes requested):
@RuaridhMacd everything looks good after the changes I made. However, there is something off with the file co2_cap_power_hsc.jl under multi-sector/genx_hsc. That file is missing definitions of some key expressions regarding RHS of constraint for the various cases. Check with main where it is correctly implemented.

RuaridhMacd commented:
- [ ] @RuaridhMacd Add warning if user activates the biofuel sector, as it's undergoing large structural changes so won't be 100% up-to-date with the energy-based hydrogen units

RuaridhMacd committed:
Update hsc_mass_2_energy.jl from the correct version in the repo

Adapted from https://github.com/RuaridhMacd/Dolphyn-HSC-Mass-2-Energy/blob/main/src/hsc_mass_2_energy.jl

RuaridhMacd committed:
Reduce significant figures on converted numbers

This will introduce slight errors in the comparison with the mass-based version

RuaridhMacd committed:
Fix naming typo in h2_truck_investment.jl

RuaridhMacd committed:
Update example inputs in sub-directories

RuaridhMacd committed:
Update co2_cap_power_hsc.jl

RuaridhMacd committed:
Add warning if user activates ModelBIO
